### ISSUE TITLE
Gracefully handle overflows

### ISSUE THREAD

**Issue by upsj** [ginkgo-project/ginkgo#1112]
There are many places where integer computations might overflow in Ginkgo. I think the most important one is the `prefix_sum` kernel. So I would like to suggest that we add an overflow guard to it, by slightly modifying the addition operation to return `-1` on overflow or if one of the operands is also `-1`. Then we can check the resulting value after the kernel has finished and throw an appropriate exception.
This gives a nicer error message than the `cuda: failed to allocate memory block of 18446744066997193496B` we usually get.

**Comment by yhmtsai** [comment]
it will give wrong error when input indeed -1?

**Comment by yhmtsai** [comment]
From prefix sum name, I would expect it allows the negative input if the type is signed type.
I would say adding the new postive_prefix_sum for the purpose, not changing the prefix_sum and adding another for the original meaning.
Is the overflow more than the GPU memory or the integer range?

**Comment by upsj** [comment]
Good point - either add a parameter `bool catch_overflows` or rename it. Though I don't agree that the original should remain, unless we really need it. This is related to arithmetic overflows, not out-of-memory issues
